### ISSUE TITLE
Upstreaming istio filters

### ISSUE THREAD

**Issue by kyessenov** [envoyproxy/envoy#29681]
Istio ships several Envoy extensions in its proxy that are not part of the Envoy distribution. This causes some friction because the release artifacts have slight differences, and adds a bunch of maintenance overhead. This issue tracks the up-streaming effort to bring these filter functionality into the core Envoy, with expected review of the API surface and the implementation.

Some notable extensions are:
* Request-based stats extension, creates metrics using the request data.
* Tunneling extensions, capture and propagation of the tunnel context into the payload context.
* Workload metadata distribution.

**Comment by pxpnetworks** [comment]
I would love to see `forward_downstream_sni ` in upstream envoy :)

**Comment by kyessenov** [comment]
`forward_downstream_sni` might be a good example for a Wasm-based extension (cc @ramaraochavali who's experimenting with this).

**Comment by kyessenov** [comment]
A few gaps in the core `jwt_authn` that require a custom add-on extension in Istio:
- changing headers should be able to flush the route cache;
- complex matching behavior for JWT payload dynamic metadata:
  - `aud` claim can be either a string or a list of strings, see https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.3
  - `scope` claim is a space separated string, see https://datatracker.ietf.org/doc/html/rfc8693#name-scope-scopes-claim

**Comment by jwendell** [comment]
@kyessenov is the eventual goal to get rid of istio/proxy repo and use envoy directly?

**Comment by kyessenov** [comment]
@jwendell Yes, that would be ideal since it aligns the interests of the projects in terms of the core data plane support, extensibility via Wasm/etc, and release artifact creation, but realistically it's unlikely to complete soon due to the amount of work needed.

**Comment by htuch** [comment]
@kyessenov status update on this one?

**Comment by kyessenov** [comment]
Update, the remaining extensions are:
- 2 metadata exchange ones (TCP and HTTP) - very unlikely to become a standard extension so should be in `contrib` IMO.
- 2 custom metrics extension (TCP and HTTP), could be ported to https://github.com/envoyproxy/envoy/issues/30619 if desired.
- 1 custom ALPN rewrite extension, could be replaced with Wasm/Lua IMO, since the logic is minimal.

Additionally, there is a suite of tests written in Go which would need to be somehow exercised in Envoy CI.

**Comment by jwendell** [comment]
This is completed, right? Should we close this issue?

**Comment by zirain** [comment]
@kyessenov @wbpcode I think this could be closed.

**Comment by jwendell** [comment]
Closing as completed, I think all the work is done now.
